src/program/latex/diagram/mod.rs

```diff
@@ -0,0 +1,176 @@
+    /// and pushed to the diagram's circuit.
+    ///
+    ///  # Arguments
+    /// `last_column` - total number of instructions from Program
+    /// `circuit` - the circuit of the diagram
+    ///
+    /// # Examples
+    /// ```
+    /// use quil_rs::{Program, program::latex::RenderSettings};
+    /// use std::str::FromStr;
+    /// let program = Program::from_str("H 0\nCNOT 0 1").expect("");
+    /// let settings = RenderSettings {
+    ///     impute_missing_qubits: true,
+    ///     ..Default::default()
+    /// };
+    /// program.to_latex(settings).expect("");
+    /// ```
+    pub(crate) fn build_wire_map(&mut self, qubits: BTreeSet<u64>) -> Result<(), LatexGenError> {
+        if qubits.is_empty() {
+            return Err(LatexGenError::RequiresQubits);
+        }
+
+        if self.settings.impute_missing_qubits {
+            let first = *qubits.first().expect("qubits is not empty");
+            let last = *qubits.last().expect("qubits is not empty");
+            for (i, qubit) in (first..last + 1).enumerate() {
+                self.wires.insert(qubit, i);
+            }
+        } else {
+            for (i, qubit) in qubits.iter().enumerate() {
+                self.wires.insert(*qubit, i);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Applies a gate from an instruction to the associated wires in the
+    /// circuit of this diagram.
+    ///
+    /// # Arguments
+    /// `gate` - the Gate of the Instruction from `to_latex`.
+    pub(crate) fn add_gate(&mut self, gate: &Gate) -> Result<(), LatexGenError> {
+        let mut cells: Vec<QuantikzCell> = vec![QuantikzCell::Empty; self.wires.len()];
+
+        let quantikz_gate = cell::QuantikzGate::try_from(gate.clone())?;
+
+        if gate.qubits.len() - quantikz_gate.ctrl_count > 1 {
+            return Err(LatexGenError::UnsupportedGate {
+                gate: gate.name.to_string(),
+            });
+        }
+
+        // parameterized non-PHASE gates are unsupported
+        if quantikz_gate.parameter.is_some() && !gate.name.contains("PHASE") {
+            return Err(LatexGenError::UnsupportedGate {
+                gate: gate.name.clone(),
+            });
+        }
+
+        let mut qubit_iterator = gate
+            .qubits
+            .iter()
+            .map(|q| {
+                q.clone()
+                    .into_fixed()
+                    .map_err(|_| LatexGenError::UnsupportedQubit)
+            })
+            .rev();
+
+        let target_qubit = qubit_iterator
+            .next()
+            .ok_or(LatexGenError::RequiresQubits)??;
+
+        let target_index = *self
+            .wires
+            .get(&target_qubit)
+            .ok_or(LatexGenError::QubitNotFound(target_qubit))?;
+
+        cells[target_index] = QuantikzCell::Gate(quantikz_gate);
+
+        for qubit in qubit_iterator {
+            let column_index = *self
+                .wires
+                .get(&qubit?)
```

Comment:
If I understand right what's happening here, I think this reads more clearly:

```suggestion
            let column_index = self
                .wires
                .get(&qubit?)
                .copied()
```